### ISSUE TITLE
Possible problem with private thread pool destruction

### ISSUE THREAD

**Issue by dfedin** [actor-framework/actor-framework#1266]
Is seems there is some problem here:
https://github.com/actor-framework/actor-framework/blob/9a09b5de89980a0fadb93b5eb13e8a280ea45f51/libcaf_core/src/actor_system.cpp#L399

Inside stop method of private_thread_pool, if we have some detached actors, private_thread_pool::running_ variable will be more than zero at this moment of time. Adding shutdown_helper node and processing it inside run_loop will do nothing since condition
https://github.com/actor-framework/actor-framework/blob/9a09b5de89980a0fadb93b5eb13e8a280ea45f51/libcaf_core/src/detail/private_thread_pool.cpp#L60
is not true - ( remaining > 0 )
Here shutdown_helper will be destroyed and lost. If some detached actor will be released after that - thread, performing stop will hang forever waiting loop thread to finish.
https://github.com/actor-framework/actor-framework/blob/9a09b5de89980a0fadb93b5eb13e8a280ea45f51/libcaf_core/src/detail/private_thread_pool.cpp#L49

Probably, some code could be added in actor_system destructor to wait until all detached actors return their threads back and stop private_thread_pool after this.

**Comment by Neverlord** [comment]
Hi! Thanks for opening the issue.

Just to understand where you are coming from: did you study the code and want point out a potential issue or is this something you have actually observed?

From my understanding, `private_threads_.stop()` gets called only after `await_all_actors_done()`. This means each detached actor must have called `scheduled_actor::cleanup` at this point, which releases the private thread:

https://github.com/actor-framework/actor-framework/blob/master/libcaf_core/src/scheduled_actor.cpp#L221

This happens *before* we decrement the running-actors count in `scheduled_actor::cleanup` (called via `super::cleanup` at the end of `scheduled_actor::cleanup`):

https://github.com/actor-framework/actor-framework/blob/master/libcaf_core/src/local_actor.cpp#L163

So when reaching `private_threads_.stop()`, it should be safe to assume that: 1) no actor (scheduled, detached, or otherwise) is still alive and 2) all private threads have been released (really a consequence of 1).

If a user spins up detached actors (or any other actor for that matter) *after* `await_all_actors_done()` already returned true, this means there must be some user-defined thread holding onto a reference to the actor system which is already out of scope. There's no reasonable strategy in this case since CAF has no knowledge of such external threads. The user should have used a shared ownership model for the actor system instead (e.g. wrapping it into a `std::shared_ptr`) to avoid the destructor from running in the first place as long as some component may still spin up workers concurrently.

AT least that is my understanding. Do I miss some code path or subtlety?

**Comment by dfedin** [comment]
Hi!

Yes, I actually observed some problems with hanging in the destructor of
actor system, joining the private_thread_pool thread.

await_all_actors_done is optional in actor system destructor here
https://github.com/actor-framework/actor-framework/blob/8f33b03ea2f5b669231b1cd354fbce79682b10a1/libcaf_core/src/actor_system.cpp#L378

Anyway, there is some problem with blocking actors even if we call
await_all_actors_done.
Blocking actors return their private thread in some specific way in
blocking_actor_running
method resume, so we can see here:
https://github.com/actor-framework/actor-framework/blob/8f33b03ea2f5b669231b1cd354fbce79682b10a1/libcaf_core/src/blocking_actor.cpp#L128
count of actors is decremented here inside cleanup *before* private threads
are returned back to the pool. So even if we will provide correct order of
thread finish and actor system destruction, we have
await_all_actors_done released,
and, probably, actor_system destructor with subsequent
private_thread_pool::stop method called *before *private thread be actually
returned to pool.

I could propose a solution here to increment actor count after private
thread release, but, probably, it will break something else.

Also, investigating hanging problem I found some inconsistency in thread
launch code inside private thread pool here:

https://github.com/actor-framework/actor-framework/blob/8f33b03ea2f5b669231b1cd354fbce79682b10a1/libcaf_core/src/detail/private_thread_pool.cpp#L68

If we will get an exception inside launch method - privated thread pool has
counter of thread already incremented which will cause the same problem in
it's stop method, loop will not break because of running_ > 0. Such
exceptions are easy to get on older Linux systems with limits on user
processes like 1024 (ulimit -u). It's actually a failure, but the process
should not hang on that.

Dmitry

пт, 4 июн. 2021 г. в 11:23, Dominik Charousset ***@***.***>:

**Comment by Neverlord** [comment]
Yeah, looks like you're right on the blocking actors. Probably the best workaround is re-implementing `launch` like this:

```cpp
void blocking_actor::launch(execution_unit*, bool, bool hide) {
  CAF_PUSH_AID_FROM_PTR(this);
  CAF_LOG_TRACE(CAF_ARG(hide));
  CAF_ASSERT(getf(is_blocking_flag));
  // Note: must *not* call register_at_system() to stop actor cleanup from
  // decrementing the count before releasing the thread.
  if (!hide) {
    [[maybe_unused]] auto count = sys.registry().inc_running();
    CAF_LOG_DEBUG("actor" << id() << "increased running count to" << count);
  }
  auto thread = sys.acquire_private_thread();
  thread->resume(new blocking_actor_runner(this, thread, hide));
}
```

By passing the `hide` flag to the `blocking_actor_runner`, we can do this instead of just calling `ctx->system().release_private_thread(thread_)`:

```cpp
  auto& sys = home_system();
  sys.release_private_thread(thread_);
  if (!hide_) {
    [[maybe_unused]] auto count = sys.registry().dec_running();
    CAF_LOG_DEBUG("actor" << id() << "decreased running count to" << count);
  }
```

> await_all_actors_done is optional in actor system destructor

Technically, yes. But there are only a few instances where setting this flag is safe. I admit that the docs don't really do that justice.

> If we will get an exception inside launch method - privated thread pool has
counter of thread already incremented which will cause the same problem

Ack. Should probably use a try-catch here for decrementing the count (and re-throwing the exception).

**Comment by Neverlord** [comment]
@dfedin could you test whether the patch to `launch` makes a difference for you?

**Comment by dfedin** [comment]
> @dfedin could you test whether the patch to `launch` makes a difference for you?

It seems it resolves a problem. But problem with exceptions on thread creation still exists.
